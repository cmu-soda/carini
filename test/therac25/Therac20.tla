----------------------------- MODULE Therac20 -----------------------------

VARIABLES state, tableState, mode, readyToFire

vars == <<state, tableState, mode, readyToFire>>
terminalVars == <<state>>
turntableVars == <<tableState>>
beamVars == <<mode, readyToFire>>

Terminal == INSTANCE Terminal
            WITH state <- state

Turntable == INSTANCE Turntable
            WITH tableState <- tableState

BeamInterlock == INSTANCE BeamInterlock
            WITH mode <- mode,
                 readyToFire <- readyToFire

Init ==
    /\ Terminal!Init
    /\ Turntable!Init
    /\ BeamInterlock!Init

TypeX ==
    \/ Terminal!TypeX /\ Turntable!TypeX /\ BeamInterlock!TypeX
TypeE ==
    \/ Terminal!TypeE /\ Turntable!TypeE /\ BeamInterlock!TypeE
TypeUp ==
    \/ Terminal!TypeUp /\ UNCHANGED turntableVars /\ UNCHANGED beamVars
TypeEnter ==
    \/ Terminal!TypeEnter /\ UNCHANGED turntableVars /\ UNCHANGED beamVars
BeamReady ==
    \/ Terminal!BeamReady /\ UNCHANGED turntableVars /\ UNCHANGED beamVars
TypeB ==
    \/ Terminal!TypeB /\ UNCHANGED turntableVars /\ BeamInterlock!TypeB
Rotate ==
    \/ UNCHANGED terminalVars /\ Turntable!Rotate /\ BeamInterlock!Rotate
fireXray ==
    \/ UNCHANGED terminalVars /\ UNCHANGED turntableVars /\ BeamInterlock!fireXray
fireElectron ==
    \/ UNCHANGED terminalVars /\ UNCHANGED turntableVars /\ BeamInterlock!fireElectron

Next ==
    \/ TypeX
    \/ TypeE
    \/ TypeUp
    \/ TypeEnter
    \/ BeamReady
    \/ TypeB
    \/ Rotate
    \/ fireXray
    \/ fireElectron

Spec == Init /\ [][Next]_vars

TypeOK ==
    /\ Terminal!TypeOK
    /\ Turntable!TypeOK
    /\ BeamInterlock!TypeOK

XrayImpliesFlatMode ==
    (mode = "xray" /\ readyToFire = TRUE) => (tableState = "flattener")

=============================================================================
