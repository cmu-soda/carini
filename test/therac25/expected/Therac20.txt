TLA+ Module: Therac20
Diff rep grouped by action:
Spec is robust against ANY behavior or environment
